### PR TITLE
[#105] Updated tpm_version to get data from TPM hardware.

### LINKED ISSUE #105: Rewrite tpm_version from tpm2 provisioner package
With the introduction of the `HIRS_ProvisionerTPM2/src/tpm_version.cpp` binary from #100, default values were provided for detecting the manufacturer and TPM version. This needs to be written to actually collect the manufacturer and version from the TPM rather than providing default values.

### PR DESCRIPTION
**Pull request by apldev4**
The TPM 2 Provisioner gets packaged with an application called
tpm_version, which reported hard-coded values for the TPM
manufacturer and version. Now it collects those things from
the TPM and reports them.

Closes #105.

### CODE REVIEW

File: HIRS_ProvisionerTPM2/docs/developer-dependencies-centos.md [hunk 1]
```diff
@@ -7,20 +7,21 @@ Please look up their respective names in the CentOS repositories.
...
+| openssl-devel     | 1.0.2k       | 1.0.2g (estimated) | CentOS 7 base         | https://github.com/openssl/openssl          |
+| protobuf-compiler | 2.5.0        | 2.4.1 (estimated)  | CentOS 7 base         | https://github.com/google/protobuf          |
+| protobuf-devel    | 2.5.0        | 2.4.1 (estimated)  | CentOS 7 base         | https://github.com/google/protobuf          |
+| re2-devel         | 20160401     | 20160201           | CentOS 7 epel-release | https://github.com/google/re2               |
+| tpm2-tss-devel    | 1.2.0        | 1.0.0              | CentOS 7 base         | https://github.com/intel/tpm2-tss           |
+| tpm2-abrmd-devel  | 1.1.0        | 1.1.0              | CentOS 7 base         | https://github.com/tpm2-software/tpm2-abrmd |
+| cmake             | 2.8.12.2     | 2.6.0 (estimated)  | CentOS 7 base         | https://cmake.org/                          |
+| cpack             | 2.8.12.2     | 2.6.0 (estimated)  | CentOS 7 base         | https://cmake.org/                          |
+| git               | 1.8.3.1      | 1.6.0 (estimated)  | CentOS 7 base         | https://github.com/git/git                  |
```
Comment (apldev6) [inline HIRS_ProvisionerTPM2/docs/developer-dependencies-centos.md]: What has changed in this file?

Comment (apldev4) [reply HIRS_ProvisionerTPM2/docs/developer-dependencies-centos.md]: In order to compile `tpm_version`, we now need to install the package `tpm2-abrmd-devel`. The whitespace on this whole table changed to get the formatting to look nice, so it looks like I changed more than I really did.

File: HIRS_ProvisionerTPM2/src/tpm_version.cpp [hunk 2]
```diff
@@ -1,11 +1,125 @@
...
+class SapiContext {
+    TSS2_TCTI_CONTEXT *tctiContext = nullptr;
+    TSS2_SYS_CONTEXT *sapiContext = nullptr;
+
+    SapiContext(TSS2_TCTI_CONTEXT* tcti_ctx, TSS2_SYS_CONTEXT* sapi_ctx)
+            : tctiContext(tcti_ctx), sapiContext(sapi_ctx) {}
+public:
+
+    ~SapiContext() {free(sapiContext); free(tctiContext);}
+    static shared_ptr<SapiContext> create() {
```
Comment (apldev2) [inline HIRS_ProvisionerTPM2/src/tpm_version.cpp]: Can you add some documentation and just some general commenting to this file?

**Comment by apldev3** [comment]
Failed at the Ubuntu packaging. Got it to package locally on my Ubuntu VM after installing the package I think it needs. Will update and push a fix.

**Comment by apldev3** [comment]
Alright the update didn't work, because the package was a) not correct and b) the correct one does not exist for Ubuntu. The reason it worked locally was because I had downloaded and built tpm2-abrmd and installed it from source, which will have to be the solution here as well.

**Comment by apldev3** [comment]
Alrighty, checked it and we're all good at this point. As expected had to build tpm2-tss and tpm2-abrmd from source on Ubuntu to get the library that the `tpm_version` binary is expecting but the Docker image is packaging it fine once again.

**Comment by apldev3** [comment]
This is all good to go @apldev4

**Comment by apldev3** [comment]
For anyone looking I updated the Dockerfiles and the Debian Developer Dependency files with the changes.

File: HIRS_ProvisionerTPM2/src/tpm_version.cpp (hunk 2 above)
Comment (apldev4) [reply HIRS_ProvisionerTPM2/src/tpm_version.cpp]: will do

File: HIRS_ProvisionerTPM2/src/tpm_version.cpp [hunk 3]
```diff
@@ -1,11 +1,190 @@
...
+/**
+ * Returns the 4-byte string represented by the bytes of value.
+ *
+ * Assumes the bytes of value are reversed.
+ * @param value the 4 bytes to be reversed and placed in the returned string
+ * @return the string represented by value
+ */
+string reversedStringValue(UINT32 value) {
+    string stringValue(4, 0);  // make a string on length 4
+    std::memcpy(&stringValue[0], &value, sizeof value);  // copy bytes in
```
Comment (apldev6) [inline HIRS_ProvisionerTPM2/src/tpm_version.cpp]: Suggest referencing the size of the target versus the source when specifying size for the memcpy.

Comment (apldev4) [reply HIRS_ProvisionerTPM2/src/tpm_version.cpp]: okay.

**Pull request merged**